## This file contains anwers Task 4.9 Part 1

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os
import gc

In [2]:
# Defining path
path = r'C:\Users\nukis\Documents\Instacart Basket Analysis'

In [3]:
# Importing 'orders_products_merged.pkl'

df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

#### Q4. Wrangle the data so that it follows consistent logic.

In [4]:
df_cust.head(10)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5   133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6   152052      Chris    Walton    Male         Montana   20    1/1/2017   
7   168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8    69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9    82820      Shawn     Chung    Male        Virginia   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374  
5             2    married   49643  
6             0     single   61746  
7             0     single   63712  
8             1    married  162432  
9             2    married   32072

In [5]:
df_cust.tail(10)

user_id First Name      Surnam  Gender           STATE  Age  \
206199   179673       Adam  Villanueva    Male         Wyoming   77   
206200   192637      Kelly    Petersen  Female         Florida   47   
206201    75529      Kathy       Avila  Female   West Virginia   25   
206202   157533        Ned       Stark    Male            Utah   53   
206203   122741       Carl       Weber    Male        Michigan   24   
206204   168073       Lisa        Case  Female  North Carolina   44   
206205    49635     Jeremy     Robbins    Male          Hawaii   62   
206206   135902      Doris    Richmond  Female        Missouri   66   
206207    81095       Rose     Rollins  Female      California   27   
206208    80148    Cynthia       Noble  Female        New York   55   

       date_joined  n_dependants        fam_status  income  
206199    4/1/2020             0  divorced/widowed  162239  
206200    4/1/2020             0            single   50638  
206201    4/1/2020             2           married   56513  
206202    4/1/2020             0            single   35973  
206203    4/1/2020             2           married   73357  
206204    4/1/2020             1           married  148828  
206205    4/1/2020             3           married  168639  
206206    4/1/2020             2           married   53374  
206207    4/1/2020             1           married   99799  
206208    4/1/2020             1           married   57095

In [6]:
df_cust.shape

(206209, 10)

In [7]:
# Renaming columns
df_cust.rename(columns= {'First Name' : 'first_name', 'Surnam' : 'surname','Gender' : 'gender','STATE' : 'state',
                          'Age' : 'age', 'n_dependants': 'dependants_number', 'fam_status' : 'marital_status'}, inplace = True)

In [8]:
df_cust.head(5)

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   dependants_number marital_status  income  
0                  3        married  165665  
1                  0         single   59285  
2                  2        married   99568  
3                  0         single   42049  
4                  1        married   40374

No columns being dropped, as they all relevant for the anaysis.

#### Q5. Complete the fundamental data quality and consistency checks.

In [9]:
# Check for missing values

df_cust.isnull().sum()

user_id                  0
first_name           11259
surname                  0
gender                   0
state                    0
age                      0
date_joined              0
dependants_number        0
marital_status           0
income                   0
dtype: int64

There are 11,259 missing values found in the 'first_name' column.

In [11]:
# View missing values

df_cust[df_cust['first_name'].isnull() == True]

user_id first_name   surname  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        dependants_number marital_status  income  
53                      2        married   41709  
73                      0         single   82518  
82                      3        married  117099  
99                      1        married  155673  
105                     3        married  151819  
...                   ...            ...     ...  
206038                  3        married   87783  
206044                  2        married  108488  
206090                  1        married   45275  
206162                  0         single   56325  
206171                  2        married   59222  

[11259 rows x 10 columns]

1. As the missing values are only 'first_name', it does not really affect our analysis. It should be considered that firstly, person name should remain confidential/privacy (unless we need it for loyalty program) and secondly, we still have acess to these user through their 'user_id' and 'surname'.
2. So, I came up with solution to fill these missing values to flag this as follows:
- If the gender is female, it will be filled with Ms.
- If the gender is male, it will be filled with Mr.

In [12]:
# Copy the DataFrame

df_cust_clean = df_cust

In [13]:
# Only as note. This command resulted in warning. Thus, it was not used.
# The next command run perfectly, no warning.

df_cust_clean['first_name'] = np.where((df_cust_clean['first_name'].isnull()) & (df_cust_clean['gender'] == 'Female'), 'Ms', df_cust_clean['first_name'])

In [14]:
df_cust_clean.loc[(df_cust_clean['first_name'].isnull()) & (df_cust_clean['gender'] == 'Female'), 'first_name'] = 'Ms'
df_cust_clean.loc[(df_cust_clean['first_name'].isnull()) & (df_cust_clean['gender'] == 'Male'), 'first_name'] = 'Mr'

In [15]:
# Check again if those above command worked.

df_cust.isnull().sum()

user_id              0
first_name           0
surname              0
gender               0
state                0
age                  0
date_joined          0
dependants_number    0
marital_status       0
income               0
dtype: int64

In [16]:
Ms = len(df_cust_clean[df_cust_clean['first_name'] == 'Ms'].value_counts(dropna = False))
Mr = len(df_cust_clean[df_cust_clean['first_name'] == 'Mr'].value_counts(dropna = False))
print(f'Ms as first_name: {Ms}')
print(f'Mr as first_name: {Mr}')
print(f'Total: {Ms + Mr}')

Ms as first_name: 5823
Mr as first_name: 5436
Total: 11259


In [17]:
# Check for mixed-type data in 'df_cust_clean'.

for col in df_cust_clean.columns.tolist():
  weird = (df_cust_clean[[col]].applymap(type) != df_cust_clean[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust_clean[weird]) > 0:
    print (col)

In [18]:
df_cust_clean.dtypes

user_id               int64
first_name           object
surname              object
gender               object
state                object
age                   int64
date_joined          object
dependants_number     int64
marital_status       object
income                int64
dtype: object

No mixed-data type is found in the 'df_cust_clean' dataframe.

In [19]:
# Check for duplicates

df_cust_clean[df_cust_clean.duplicated()]

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_joined, dependants_number, marital_status, income]
Index: []

No duplicates are found in the 'df_cust_clean' dataframe.

In [20]:
# Export the clean customer DataFrame to avoid Jupyter crash.

df_cust_clean.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'customer_cleaned.pkl'))

In [21]:
gc.collect()

0

#### Q6. Combine your customer data with the rest of your prepared Instacart data.

As my Jupyter kept crashing, I divided this notebook into 3 parts. This notebook will be continued to Part1B and Part1C.